[PLT-5278] shows generic thumbnail from storage when max image exceeded

johncoleman83 opened the pull request:
#### Summary
This change allows the same generic thumbnail image from storage to be used as the thumbnail path for when a file upload image dimensions are exceeded.  If the generic thumbnail image is not yet in storage, then it is added from the client package path: `mattermost/client/images/image.png`.  Therefore, in order to have that `image.png` accessible, this change has a dependency, which is the file `webpack.config.js` modification from the `mattermost-webapp`.  The Pull Request that includes the dependency on the mattermost-webapp can be found here: [pull/286](https://github.com/mattermost/mattermost-webapp/pull/286)

**Work in Progress**
This is a work in progress because I have not yet added tests.  I wanted to open up this code and approach for consultation from the maintainers/ developers.  Another comment is that the generic thumbnail is never added to the app store, and I was not sure if is a change that should be added.

#### Ticket Link
[[Help Wanted] [PLT-5278]](https://github.com/mattermost/mattermost-server/issues/5485)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] Added or updated unit tests (required for all new features)
- [x] Added API documentation (required for all new APIs)
- [x] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json) updates
- [x] Touches critical sections of the codebase (auth, upgrade, etc.)

saturninoabril commented on app/file.go:
```diff
@@ -417,6 +424,25 @@ func getImageOrientation(input io.Reader) (int, error) {
 	}
 }
 
+func genericThumbnailHandler() (*model.AppError) {
```
I don't think this is necessary.

johncoleman83 replied:
I did not know that Mattermost app server could find an image that was located in a path outside of the storage model.  This function was to add the thumbnail image to the storage model so that it would be made accessible to any requests to `info.ThumbnailPath`, which = `GENERIC_THUMBNAIL_PATH`.

saturninoabril replied:
As commented above regarding path of `generic_thumbnail_image.png` but @hmhealey might give better approach to this.

saturninoabril commented on app/file.go:
```diff
@@ -55,6 +56,9 @@ const (
 	IMAGE_THUMBNAIL_PIXEL_HEIGHT = 100
 	IMAGE_PREVIEW_PIXEL_WIDTH    = 1024
+
+	GENERIC_THUMBNAIL_LOCAL_PATH   = "./client/images/image.png"
```
Maybe this is enough:
```
GENERIC_THUMBNAIL_PATH   = "client/images/generic_thumbnail_image.png"
```

saturninoabril commented on app/file.go:
```diff
@@ -341,7 +346,9 @@ func HandleImages(previewPathList []string, thumbnailPathList []string, fileData
 			go func(img *image.Image, path string, width int, height int) {
 				defer wg.Done()
-				generateThumbnailImage(*img, path, width, height)
+				if thumbnailPathList[i] != GENERIC_THUMBNAIL_STORAGE_PATH {
```
Could be like:
```
if thumbnailPathList[i] != GENERIC_THUMBNAIL_PATH {
```

saturninoabril commented on app/file.go:
```diff
@@ -310,21 +314,22 @@ func (a *App) DoUploadFile(now time.Time, rawTeamId string, rawChannelId string,
 	info.Path = pathPrefix + filename
 
 	if info.IsImage() {
```
I'm thinking something like this:
```
	if info.IsImage() {
		nameWithoutExtension := filename[:strings.LastIndex(filename, ".")]
		// Check dimensions before loading the whole thing into memory later on
		if info.Size > MaxImageSize {
			err := model.NewAppError("uploadFile", "api.file.upload_file.large_image.app_error", map[string]interface{}{"Filename": filename}, "", http.StatusBadRequest)
			return nil, err
		} else if info.Width*info.Height > MaxImageSize {
			info.ThumbnailPath = GENERIC_THUMBNAIL_PATH
		} else {
			info.ThumbnailPath = pathPrefix + nameWithoutExtension + "_thumb.jpg"
		}

		info.PreviewPath = pathPrefix + nameWithoutExtension + "_preview.jpg"
	}
```

johncoleman83 replied:
Hi @saturninoabril!  Thanks for the review.  If the path stored for the thumbnail image is: `client/images/generic_thumbnail_image.png`.  Does this mean that any request for: `info.ThumbnailPath = GENERIC_THUMBNAIL_PATH` would not be made to storage, but instead to the app server's `client` directory?  I thought that by default, requests to info.ThumbnailPath` were made to storage, this is why I added the thumbnail to storage.  If you want to make the path on the app server side instead of storage, is there any necessary logic to be added, or will this function as you hope?

johncoleman83 replied:
Also, @saturninoabril, the first `if` conditional you added: `info.Size > MaxImageSize`, I believe this check already occurs here in [`api4/file.go`](https://github.com/mattermost/mattermost-server/blob/10c5a927cb619f1aa2a599cbe3667820f4766bda/api4/file.go#L67).  Did you still want to run this check a second time?

saturninoabril replied:
1) Path of `generic_thumbnail_image.png`, I don't think it should reside to local or S3 storage. If file's thumbnail is equal to `GENERIC_THUMBNAIL_PATH` then it should not hit the `files/{file_id:[A-Za-z0-9]+}/thumbnail` api. The webapp (or client) should immediately render `generic_thumbnail_image.png`.
But @hmhealey might have a better approach to this.
2) You're correct. I missed that.

hmhealey commented:
Unfortunately, we can't rely on knowing where that image is. GitLab, for example, bundles us with the client located in an entirely different directory, and we've discussed having the web app served from a different server so we may not even be able to assume that the file we need is on the same server.

Either that, or instead of the "no thumbnail" response being a 400 error, we instead redirect to a url that we know will contain a generic thumbnail like `/static/images/icons/image.png`. We'll need to make a change to [webpack.config.js](https://github.com/mattermost/mattermost-webapp/blob/master/webpack.config.js#L319) to have it actually store the generic thumbnails there since right now they're hashed and stored by webpack, but that would mean the client doesn't have to care.

I was also thinking we could have the client know how to handle that error and render a generic thumbnail from there, but that may cause issues for outdated clients, particularly the mobile app. Having the server redirect is also probably easier to implement since it's mostly server changes.

johncoleman83 commented:
Hi @hmhealey and @saturninoabril.  Thanks sincerely for your insights.  I think it is a great point (and nice option) to have the client package hosted in a different path or server from the app backend server.  With that in mind and your other suggestions, I think that these updates in the code will be a good option for this Help Topic of using a generic thumbnail image for images that exceed the `MaxImageSize`.

**To explain the updates:**

- I have rebased all newest changes in `app/file.go` and incorporated the changes in my updates.
- I added the `generic_thumbnail_image.png` to the package via a new directory `./images`, which is copied in the package from the `build/release.mk`.
- This image, `generic_thumbnail_image.png`, will only be used in the event that the generic thumbnail is not found in the designated backend file storage (i.e. S3 or backend files path).
  - this check occurs via a new function: `func (a *App) genericThumbnailHandler() (*model.AppError)`, which is called when uploading a new image that meets the criteria of having dimensions greater than: `MaxImageSize`.
  - the function `genericThumbnailHandler()` checks if the generic image is in the backend storage, if not there, it is added.
- Once the generic thumbnail image is added to the designated backend storage (i.e. S3 or backend files path), all images with dimensions that exceed the `MaxImageSize`, will use this path as the path for the thumbnail image.
- no thumbnail is generated since it already exists for all subsequent uploaded images that meet the criteria of exceeding the `MaxImageSize`.
- tests pass on my end.

saturninoabril commented on app/file.go:
```diff
@@ -453,6 +461,25 @@ func getImageOrientation(input io.Reader) (int, error) {
 	}
 }
 
+func (a *App) genericThumbnailHandler() (*model.AppError) {
```
This failed on style check. Should change to 
```
func (a *App) genericThumbnailHandler() *model.AppError {
```

johncoleman83 replied:
Thanks, I removed the parentheses.

saturninoabril commented:
Also, I've noticed that you're pushing your master (`johncoleman83:master`). Better to create a branch and then submit a PR against `mattermost:master`.

`git checkout -b GH-5485`

johncoleman83 commented:
Hi @saturninoabril, thanks for the tip on using branches.  I see how to modify the `base` branch in an open pull request, but I do not know of a way to modify the `from` branch in an opened pull request.  If you know how to do this and have permission to, the new branch from my repo is `PLT-5278`.  Otherwise, would you like me to create a new Pull Request from this new branch on my repo?

hmhealey commented:
Hi @johncoleman83, I talked it over with Saturn and some other members of the team, and we'd rather not write the generic thumbnail to disk for these images since it will be adding a lot of coupling between what we store on the server and what we display in the UI which will make future changes to the UI very difficult. For example, we have an in-progress design (it's [here](https://mattermost.atlassian.net/browse/UX-26) if you're interested in seeing it) that will change how post attachments are displayed, and it'll include a set of new generic icons, so having those written to disk would make it very difficult to update those.

Now, we also discussed a bit about how we'd like this done so that it'll be easiest to implement now, and to change in the future. If you're on our pre-release server, the discussion is [here](https://pre-release.mattermost.com/core/pl/f4aseurosbr4mkwwg1455c94iw). The conclusion we've come to is that it would work best to have the client implement the handling for this since it should be relatively easy, and it'll allow us the most freedom in the future when we do redesign the UI for this. This'll mean that
- The server changes will be minimal since they just need to allow the file to be uploaded without generating a thumbnail or preview
- The web app will need to know to properly handle when the server returns that there's no thumbnail/preview image for so that it can fall back to the generic thumbnail or display the generic preview
- The mobile app already handles errors, but it tries to download the thumbnail a few times before giving up, so we should have it bypass that behaviour when it sees the "no thumbnail" error. If you're not wanting to work on that (understandable since React Native is a pain to set up at times), one of us can take a look at this part.

If you have any questions or need any other help with this, feel free to let one of us know on here or on https://pre-release.mattermost.com

crspeller commented:
@johncoleman83 @hmhealey Going to close this because of inactivity. Feel free to reopen when a new PR is ready.
